proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,432 @@
+```
+
+Valid resource types for the texture handle parameter and the corresponding
+ active coordinates and offsets are the same as for the `sampleBias` operation
+ except that `TextureCubeArray` resources are invalid for `sampleCmpGrad`.
+The sampler handle parameter must be of type `SamplerComparisonState`.
+
+The existing `dx.op.calculateLOD` intrinsic is reused for the new
+ `CalculateLevelOfDetail` overloads taking `SamplerComparisonState` parameters.
+
+### SPIR-V Additions
+
+To implement the new `SampleCmpGrad` and `SampleCmpBias` texture methods,
+ the existing `OpImage*SampleDrefImplicitLod` operations can be used with the
+ optional image operands set to the value for `Grad`(0x4) or `Bias`(0x1) set
+ to the values provided by the user.
+
+To implement the new `CalculateLevelOfDetail[Unclamped]` overloads,
+ the existing `OpImageQueryLod` operation may be used with comparison state images.
+
+### Diagnostic Changes
+
+#### Removed errors
+
+* Using `CalculateLevelOfDetail[Unclamped]` with a `SamplerComparisonState`
+  sampler will no longer produce an error when compiling with appropriate shader
+  model versions.
+* Invoking new methods `SampleCmpBias` and `SampleCmpGrad` with valid texture
+  objects and parameters will not produce an unrecognized function error.
+
+#### New Errors
+
+* Using `SampleCmpBias` in code reachable from the entry point of unsupported
+  shader stages will produce an error.
+* Using `SampleCmpBias`, `SampleCmpGrad`, or the new `CalculateLevelOfDetail*`
+  overloads in an unsupported shader model version will produce an special error
+  indicating that a later shader model version is required.
+
+#### Validation Changes
+
+* DXIL validation of shaders using this feature requiring that
+  `@dx.op.calculateLOD` have a non-comparison sampler handle are removed.
+* DXIL validation of LOD Bias range must be performed for `dx.op.sampleCmpBias`.
+  (see Instr.ImmBiasForSampleB)
+* DXIL validation of allowed resource types must be performed for
+  `dx.op.sampleCmpBias` and `dx.op.sampleCmpGrad`.
+  (See Instr.SampleCompType, Instr.SamplerModeForSampleC,
+   Instr.ResourceKindForCalcLOD, Instr.ResourceKindForSample, and
+   Instr.ResourceClassForSamplerGather)
+
+### Runtime Additions
+
+#### Runtime information
+
+A shader flag indicating usage of `SampleCmpBias` or `SampleCmpGrad` to indicate
+ that the shader requires the presence of a capability bit indicating support.
+
+#### Device Capability
+
+Applications can query the availability
+ of these features by
+ passing `D3D12_FEATURE_D3D12_OPTIONS22`
+ as the `Feature` parameter
+ and retrieving the `pFeatureSupportData` parameter
+ as a struct of type `D3D12_FEATURE_DATA_D3D12_OPTIONS22`.
+The relevant parts of these structs are defined below.
+
+```C++
+typedef enum D3D12_FEATURE {
+    ...
+    D3D12_FEATURE_D3D12_OPTIONS22
+} D3D12_FEATURE;
+
+typedef struct D3D12_FEATURE_DATA_D3D12_OPTIONS22 {
+    ...
+    BOOL ExtendedComparisonSamplesSupported;
```

Comment:
Could `Samples` be replaced with `Sampling` to match the spec?  Also, for some reason "Sampling" seems less awkward to me.

Reply:
Looks like D3D went with `SampleCmpGradientAndBiasSupported` and OPTIONS21. So I'll change that.

Reply:
Done